src/ast.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 VMware, Inc.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+use serde_json::Value;
+
+/// A path is a tree of selector nodes.
+///
+/// For example, the JSONPath `$.foo.bar` yields this AST:
+///
+/// ```text
+///               ^
+///              / \
+///             ^   \___ DotName("bar")
+///            / \
+///           ^   \___ DotName("foo")
+///          /
+/// Root ___/
+/// ```
+///
+/// A more complicated example: `$.foo[1,2]["bar"]`:
+///
+/// ```text
+///                  ^
+///                 / \
+///                ^   \___ Union
+///               / \            \
+///              ^   \___ Union   \
+///             /            \     [Field("bar")]
+///            ^              \
+///           / \              [Number(1), Number(2)]
+///          /   \
+/// Root ___/     \___ DotName("foo")
```

Comment:
The two diagrams still seem inconsistent. I'd expect the part of the tree nearest the root to be identical since `$.foo` is a common prefix of both selectors.

Reply:
That's because the expressions have right to left precedence

Reply:
Still don't get it. `$.foo.bar` produces the AST:
```
Sel(Sel(Root, DotName("foo")), DotName("bar"))
```
and `$.foo[1,2]["bar"]` produces:
```
Sel(Sel(Sel(Root, DotName("foo")), Union([Number(1), Number(2)])), Union([Field("bar")]))
```
`Root` appears inside `Sel(Root, DotName("foo"))` in both cases. Shouldn't this subtree look the same in both diagrams?

Reply:
Uh :brainfart:; fixed